plugins/languages/treesitter/treesitter.nix

```diff
@@ -6,210 +6,317 @@
+        })
+        [
+          "gccPackage"
+          "nodejsPackage"
+          "treesitterPackage"
+        ];
 
-        incremental_selection =
-          if cfg.incrementalSelection.enable then
-            {
-              enable = true;
-              keymaps = {
-                init_selection = cfg.incrementalSelection.keymaps.initSelection;
-                node_incremental = cfg.incrementalSelection.keymaps.nodeIncremental;
-                scope_incremental = cfg.incrementalSelection.keymaps.scopeIncremental;
-                node_decremental = cfg.incrementalSelection.keymaps.nodeDecremental;
-              };
-            }
-          else
-            null;
-
-        indent = if cfg.indent then { enable = true; } else null;
-
-        ensure_installed = if cfg.nixGrammars then [ ] else cfg.ensureInstalled;
-        ignore_install = cfg.ignoreInstall;
-        parser_install_dir = cfg.parserInstallDir;
-      } // cfg.moduleConfig;
-    in
-    mkIf cfg.enable {
-      extraConfigLua =
-        (optionalString (cfg.parserInstallDir != null) ''
-          vim.opt.runtimepath:append("${cfg.parserInstallDir}")
+    extraConfigLua =
+      ''
+        do
+      ''
+      +
+        # NOTE: Upstream state that the parser MUST be at the beginning of runtimepath.
+        # Otherwise the parsers from Neovim takes precedent, which may be incompatible with some queries.
+        (optionalString (cfg.settings.parser_install_dir != null) ''
+          vim.opt.runtimepath:prepend("${cfg.settings.parser_install_dir}")
         '')
-        + ''
-          require('nvim-treesitter.configs').setup(${helpers.toLuaObject tsOptions})
-        ''
-        + (optionalString (cfg.languageRegister != { }) ''
-          __parserFiletypeMappings = ${helpers.toLuaObject cfg.languageRegister}
-
-          for parser_name, ft in pairs(__parserFiletypeMappings) do
-            require('vim.treesitter.language').register(parser_name, ft)
-          end
-        '');
-
-      extraFiles = mkIf cfg.nixvimInjections {
-        "queries/nix/injections.scm".text = ''
-          ;; extends
-
-          (binding
-            attrpath: (attrpath (identifier) @_path)
-            expression: [
-              (string_expression (string_fragment) @lua)
-              (indented_string_expression (string_fragment) @lua)
-            ]
-            (#match? @_path "^extraConfigLua(Pre|Post)?$"))
-
-          (binding
-            attrpath: (attrpath (identifier) @_path)
-            expression: [
-              (string_expression (string_fragment) @vim)
-              (indented_string_expression (string_fragment) @vim)
-            ]
-            (#match? @_path "^extraConfigVim(Pre|Post)?$"))
-        '';
-      };
+      + ''
+        require('nvim-treesitter.configs').setup(${helpers.toLuaObject cfg.settings})
+      ''
+      + (optionalString (cfg.languageRegister != { }) ''
+        local __parserFiletypeMappings = ${helpers.toLuaObject cfg.languageRegister}
```

Comment:
Sorry to change my mind on this, but I think the `do .. end` scope only needs to wrap this optional string:

```nix
(optionalString (cfg.languageRegister != { }) ''
  do
    local __parserFiletypeMappings = ${helpers.toLuaObject cfg.languageRegister}

    for parser_name, ft in pairs(__parserFiletypeMappings) do
      require('vim.treesitter.language').register(parser_name, ft)
    end
  end
'')
```

Because the only variable that could leak to the rest of the init.lua scope is `__parserFiletypeMappings`